src/abstractdataframe/reshape.jl

```diff
@@ -185,39 +193,72 @@ long = stack(wide)
-    refkeycol = CategoricalArray{Union{eltype(df[rowkey]), Missing}}(df[rowkey])
-    # make sure we report only levels actually present in rowkey column
-    # this is consistent with how the other unstack method based on groupby works
+    refkeycol = copy(CategoricalArray{Union{eltype(df[rowkey]), Missing}}(df[rowkey]))
     droplevels!(refkeycol)
-    valuecol = df[value]
     keycol = CategoricalArray{Union{eltype(df[colkey]), Missing}}(df[colkey])
+    valuecol = df[value]
+    Nrow = length(refkeycol.pool)
+    Ncol = length(keycol.pool)
+    df2m = [similar_missing(valuecol, Nrow) for i in 1:Ncol]
+    _unstack(df, rowkey, colkey, value, keycol, valuecol, df2m, refkeycol)
+end
+
+function _unstack(df::AbstractDataFrame, rowkey::Int,
+                  colkey::Int, value::Int, keycol, valuecol, df2m, refkeycol)
     Nrow = length(refkeycol.pool)
     Ncol = length(keycol.pool)
-    payload = DataFrame(Any[similar_missing(valuecol, Nrow) for i in 1:Ncol], map(Symbol, levels(keycol)))
-    nowarning = true
+    hadmissing = false # have we encountered missing in refkeycol
+    mask_filled = falses(Ncol, Nrow+1) # has a given [col,row] entry been filled?
+    warned_missing_1 = false # do we print duplicate entries warning
+    warned_missing_2 = false # do we print missing in keycol
+    keycol_order = Vector{Int}(CategoricalArrays.order(keycol.pool))
+    refkeycol_order = Vector{Int}(CategoricalArrays.order(refkeycol.pool))
     for k in 1:nrow(df)
-        j = Int(CategoricalArrays.order(keycol.pool)[keycol.refs[k]])
-        i = Int(CategoricalArrays.order(refkeycol.pool)[refkeycol.refs[k]])
-        if i > 0 && j > 0
-            if nowarning && !ismissing(payload[j][i])
-                warn("Duplicate entries in unstack.")
-                nowarning = false
+        kref = keycol.refs[k]
+        if kref <= 0 # we have found missing in colkey
+            if !warned_missing_2
+                warn("Missing value in '$(_names(df)[colkey])' variable at row $k. Skipping.")
+                warned_missing_2 = true
+            end
+            continue # skip processing it
+        end
+        j = keycol_order[kref]
+        refkref = refkeycol.refs[k]
+        if refkref <= 0 # we have found missing in rowkey
+            if !hadmissing # if it is the first time we have to add a new row
+                hadmissing = true
+                # we use the fact that missing is greater than anything
+                for i in eachindex(df2m)
+                    push!(df2m[i], missing)
+                end
             end
-            payload[j][i]  = valuecol[k]
+            i = length(df2m[1])
+        else
+            i = refkeycol_order[refkref]
+        end
+        if (!warned_missing_1) && mask_filled[j, i]
+            warn("Duplicate entries in unstack at row $k.")
+            warned_missing_1 = true
         end
+        df2m[j][i] = valuecol[k]
+        mask_filled[j, i] = true
```

Comment:
`[j, i]` is less natural than `[i, j]`, is this to optimize memory access patterns? Do we expect the data to be sorted (and therefore filled) by rows or by columns? Given the layout of data frames, it would be optimal to fill one column at a time. I admit I don't really understand whether that's the case currently or not.


We should also ensure that's consistent with what `stack` outputs.

Reply:
`stack` sorts by variable not by index. So if we `unstack` something that was stacked `j` is changing faster than `i` so I guess this layout is better. Yes?

Reply:
If that's the case (haven't checked), then it's not optimal: we need `i` to change the fastest, since accessing one column at a time is more efficient for data frames. Maybe we should change `stack`? Though we should check what R/dplyr/Pandas do first.

Reply:
Two points (contradictory though):
* I was wrong we fill the data column-by-column (so `i` is changing faster)
* However, both - result of `stack` and then random permutation of result of `stack` are minimally faster when we have the current order (i.e. `mask_filled = falses(Ncol, Nrow)`). I do not understand why.

Anyway this does not matter much - it is again the case of minimal difference in performance. So I would leave it as is.

Reply:
OK, thanks for checking. That's really weird, but well... For maximal performance, we could also use an `Array{Bool}`, but probably not worth the increased memory use.

Reply:
BTW, have you tried adding `@inbounds` before indexing expressions where we are sure indices are correct?

Reply:
the code is complex and easy to make mistake in reasoning. I would leave out `@inbounds` until it is well tested by users. I will add a TODO in the comment.

Reply:
No need for a TODO: either you know it's OK or you don't.

At least it should be fine when you index `refkeycol` and `keycol`.

Reply:
I know 😄, but conditional on what you have written, so I would not risk. I have checked that the gains form `@inbounds` are minimal <1%.